### PR TITLE
Fixed base model class name extraction from PeftModels

### LINKED ISSUE #27161: Trainer doesn't shift labels for CAUSAL_LM PEFT models with label smoothing enabled
### System Info

- `transformers` version: 4.35.0.dev0
- Platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.2
- Accelerate version: 0.24.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Tensorflow version (GPU?): 2.14.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: RTX 3090
- Using distributed or parallel set-up in script?: No

### Who can help?

@pacman100, @muellerzr 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I am not sure how to articulate this silent bug with a code snippet. I will try to explain it referring to the code and hopefully it will be clear.

The [`compute_loss`](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2672) function in the [Trainer](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L229) class gets the `model_name` from `model.base_model` when the model is a `PeftModel` as written [here](https://github.com/huggingface/transformers/blob/84724efd101af52ed3d6af878e41ff8fd651a9cc/src/transformers/trainer.py#L2690). The `base_model` of a `PeftModel` is defined [here](https://github.com/huggingface/peft/blob/884b1ac3a8ef49c9301b5bbf02e8bc64349e95f9/src/peft/peft_model.py#L119) as one of [`PEFT_TYPE_TO_MODEL_MAPPING`](https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L68). The 'true' base model is actually stored at `base_model.model` as declared [here](https://github.com/huggingface/peft/blob/884b1ac3a8ef49c9301b5bbf02e8bc64349e95f9/src/peft/tuners/tuners_utils.py#L70).

The issue is--in [this](https://github.com/huggingface/transformers/blob/84724efd101af52ed3d6af878e41ff8fd651a9cc/src/transformers/trainer.py#L2693) line of `compute_loss` method, the check to shift labels is done by seeing if `model_name` is inside the [`MODEL_FOR_CAUSAL_LM_MAPPING_NAMES`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/auto/modeling_auto.py#L384) list. Since the `model_name` for a `PeftModel` isn't in that list, the labels aren't shifted.

From what I can tell, a simple fix without breaking anything else could be to modify [this](https://github.com/huggingface/transformers/blob/84724efd101af52ed3d6af878e41ff8fd651a9cc/src/transformers/trainer.py#L2690) line to:

```
model_name = unwrap_model(model.base_model.model)._get_name()
```

### Expected behavior

The labels should be shifted for causal language modelling tasks even when using peft models.

### PR DESCRIPTION
**Pull request by kkteru**
# What does this PR do?

Fixes #27161

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@pacman100, @muellerzr

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @younesbelkada

File: src/transformers/trainer.py
Comment (younesbelkada) [inline src/transformers/trainer.py:2689]:
```suggestion
            unwrapped_model = unwrap_model(model)
            if is_peft_available() and isinstance(unwrapped_model, PeftModel):
                model_name = unwrapped_model.base_model.model._get_name()
            else:
                model_name = unwrapped_model._get_name()
            if model_name in MODEL_FOR_CAUSAL_LM_MAPPING_NAMES.values():
                loss = self.label_smoother(outputs, labels, shift_labels=True)
            else:
                loss = self.label_smoother(outputs, labels)
            del unwrapped_model
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot! 
I tried to reproduce and it is indeed a bug - however your fix does not properly take care of the case where users do DDP + PEFT. I propose to create first a dummy variable `unwrapped_model` that simply unwraps the model in case it is DDP or FSDP then perform the checks you suggested. What do you think?

File: src/transformers/trainer.py
Comment (younesbelkada) [reply src/transformers/trainer.py:2689]: Don't accept this suggestion otherwise it will create a weird diff but I did this so that you can see what I meant

**Review by younesbelkada — APPROVED** [review]
Thanks for your great work on this!

File: src/transformers/trainer.py
Comment (amyeroberts) [inline src/transformers/trainer.py:2691]: For my own understanding, was `unwrap_model(model.base_model)._get_name()` ever working? 

I'm asking to understand whether we would still need the case 
`model_name = unwrapped_model.base_model._get_name()`

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this! 

Just a small question - once clarified I think we're good to merge :)

File: src/transformers/trainer.py
Comment (younesbelkada) [reply src/transformers/trainer.py:2691]: > was unwrap_model(model.base_model)._get_name() ever working?

If `model` is a DistrbutedDataParallel or a FSDP wrapped module `model.base_model` would fail (you need to unwrap `model` first) + `unwrap_model(model.base_model)._get_name()` would also fail because the model is stored in model.base_model.model. Let me know if you want more clarifications

Comment (amyeroberts) [reply src/transformers/trainer.py:2691]: OK, but what if the model isn't distributed? Is it possible for those models to be hitting this logic branch?

Comment (kkteru) [reply src/transformers/trainer.py:2691]: If the model isn't distributed and it is a PeftModel, `unwrap_model(model.base_model)._get_name()` would not fail but return the wrong `model_name` due to the reason mentioned in #27161.

In other words, without the DDP/FSDP the fix of just replacing `model.base_model` with `model.base_model.model` would have worked. IMO, the change @younesbelkada suggested fixes a different bug that I did not explicitly mention in the original issue, i.e., PeftModel with DDP/FSDP + label smoothing would throw an error.

Comment (amyeroberts) [reply src/transformers/trainer.py:2691]: Thanks both for your explanations. Final question: should this be applied to the other places in trainer where there's similar logic e.g. [here](https://github.com/huggingface/transformers/blob/00d8502b7ade5aa3da43b13f23bb447faa6d459e/src/transformers/trainer.py#L649)

Comment (younesbelkada) [reply src/transformers/trainer.py:2691]: But we can update it for consistency, what do you think?

Comment (kkteru) [reply src/transformers/trainer.py:2691]: That sounds like a good idea. Let me try to do that, if that is okay. I see two potential changes in the neft activate and deactivate functions. I will try to do a more thorough scan and push another commit for review.

@younesbelkada Great work on integrating NEFT so quick, thank you so much!

Comment (younesbelkada) [reply src/transformers/trainer.py:2691]: sounds great, thank you @kkteru !

Comment (kkteru) [reply src/transformers/trainer.py:2757]: Just pushed the changes. I don't think there is any other place where this needs to be changed. One place that came close was [this](fix-trainer-compute-loss-with-peft-non-zero-label-smoothing) prefix definition when loading the adapter state_dict, but I think that was correctly declared.

**Review by younesbelkada — APPROVED** [review]
Thanks again for your great work!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating!

File: src/transformers/trainer.py
Comment (kkteru) [reply src/transformers/trainer.py:2757]: > If `model` is a DistrbutedDataParallel or a FSDP wrapped module `model.base_model` would fail (you need to unwrap `model` first) + `unwrap_model(model.base_model)._get_name()` would also fail because the model is stored in model.base_model.model. Let me know if you want more clarifications

I actually noticed similar issue in the sft_trainer/neft support of trl package [here](https://github.com/huggingface/trl/blob/cc1de9820a2375bd1a1a69ab2758b139cd7b1d69/trl/trainer/sft_trainer.py#L286) and [here](https://github.com/huggingface/trl/blob/cc1de9820a2375bd1a1a69ab2758b139cd7b1d69/trl/trainer/sft_trainer.py#L384). Happy to push a PR there or leave it to you to clean up after.

Comment (younesbelkada) [reply src/transformers/trainer.py:2757]: ah yes, if you could submit PRs on TRL that would be great as well ! Thanks @kkteru !

**Pull request merged**
